subworkflows/local/utils_nfcore_mcmicro_pipeline/main.nf

```diff
@@ -181,60 +185,166 @@ def validateInputParameters() {
 //
 // Validate channels from input samplesheet
 //
-def validateInputSamplesheet(input) {
-    // TODO: Add sample sheet validation.
-    return input
+
+def validateInputMarkersheet( sheet_data ) {
+
+    def idx_marker_name = input_sheet_index("marker", "marker_name")
+    def idx_channel_number = input_sheet_index("marker", "channel_number")
+    def idx_cycle_number = input_sheet_index("marker", "cycle_number")
+    def marker_name_list = []
+    def channel_number_list = []
+    def cycle_number_list = []
+    def outstr = ""
+
+    sheet_data.each { curr_list ->
+        outstr = outstr + curr_list.join(",") + '\n'
+        def idx = 0
+        curr_list.each { curr_val ->
+            curr_pair = [-1, -1]
+            if (idx == idx_marker_name) {
+                if (marker_name_list.contains(curr_val)) {
+                    error("Error: duplicate marker name found in marker sheet!")
+                } else {
+                    marker_name_list.add(curr_val)
+                }
+            } else if (idx == idx_channel_number) {
+                if (curr_val <= 0) {
+                    error("Error: channel_number must be >= 1")
+                } else if (channel_number_list && (curr_val != channel_number_list[-1] && curr_val != channel_number_list[-1].toInteger() + 1)) {
+                    error("Error: channel_number cannot skip values and must be in order!")
+                } else {
+                    channel_number_list.add(curr_val)
+                }
+            } else if (idx == idx_cycle_number) {
+                if (curr_val <= 0) {
+                    error("Error: cycle_number must be >= 1")
+                } else if (cycle_number_list && (curr_val != cycle_number_list[-1] && curr_val != cycle_number_list[-1].toInteger() + 1)) {
+                    error("Error: cycle_number cannot skip values and must be in order!")
+                } else {
+                    cycle_number_list.add(curr_val)
+                }
+            }
+            idx++
+        }
+    }
+
+    // uniqueness of (channel, cycle) tuple in marker sheet
+    def test_tuples = [channel_number_list, cycle_number_list].transpose()
+    if ( test_tuples.size() != test_tuples.unique( false ).size() ) {
```

Comment:
This is OK, but it would be nicer to report the actual duplicate pair(s):
```groovy
def dups = test_tuples.countBy{ it }.findAll{ _, count -> count > 1 }*.key
if (dups) {
    error("Error: duplicate [channel, cycle] pairs: ${dups}")
}
```

Reply:
Nice, your code is much better and you're right, much better to also report the duplicate(s).  I've replaced my code with what you have here.  Thanks!